### PR TITLE
Add populate exchange

### PR DESCRIPTION
**Pull request by andyrichardson**
For discussion, see https://github.com/FormidableLabs/urql/pull/471

### Todo
 - [x] Use helper fns
 - [x] Add export from entrypoint

### CODE REVIEW

**Comment by JoviDeCroock** [comment]
> The console.warn and throw Error can be replaced with our warning and invariant helpers and there's probably some other stuff that can be reused 💯

Then we do need to add codes for them

**Comment by andyrichardson** [comment]
Types look pretty vague on Invariant (i.e. what is code?) - let me know if I'm doing this wrong.

Also, would be nice to do 

```ts
if (condition)
  throw invariant(...args)
}

someOtherCode();
```

rather than

```ts
if (condition) {
  invariant(...args); // Intention to throw isn't clear to newcomers
}

someOtherCode();
```

**Comment by andyrichardson** [comment]
> Let's do this! 💯 Great work Andy!
> 
> We still need to add the export though

Done :+1:

**Comment by andyrichardson** [comment]
Looked at typing the errors but I'm not sure numbers is the way to do this.

String literals would make more sense:

Numbers
```ts
// Types
type ErrorCode = InvalidGraphQLDocumentCode | InvalidCacheCallCode;

type InvalidGraphQLDocumentCode = 1;

type InvalidCacheCallCode = 2;

// Usage
warn(msg, 1); // 1 ???
```

Literals
```ts
// Types
type ErrorCode = "INVALID_GRAPHQL_DOCUMENT" | "INVALID_CACHE_CALL";

// Usage
warn(msg, "INVALID_GRAPHQL_DOCUMENT"); // makes more sense
```

**Comment by kitten** [comment]
@andyrichardson They're not any typed/relevant type information, but they're just numeric identifiers that we add to our `help.md` to add more information on how a warning/error can be resolved and why it happened 🆘  https://github.com/FormidableLabs/urql-exchange-graphcache/blob/master/docs/help.md

So typically just choose the next highest number in that doc and add a new section and copy that info over 😅 It's not linted (yet)

**Comment by andyrichardson** [comment]
> They're not any typed/relevant type information, but they're just numeric identifiers that we add to our help.md to add more information on how a warning/error can be resolved and why it happened

@kitten yeah I totally get what the numbers are for. Literals would do the same job, just a better DX.

Going back on topic, what's left for this PR? The only thing I'm aware of is the number being thrown from the invariant call. Can't see anyone looking to wrap the exchange in a try/catch so I wouldn't put too much thought into it.

File: src/populate.ts [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+  const type =
+    isListType(typeInfoType) || isNonNullType(typeInfoType)
+      ? typeInfoType.ofType
+      : typeInfoType;
+
+  if (isAbstractType(type)) {
+    invariant(
+      'PopulateExchange',
+      'Invalid TypeInfo state. "getTypeName" does not expect to receive an abstract type.',
+      1
```
Comment (kitten) [inline src/populate.ts]: Still an error code that isn't in `help.md` but I'll fix that up in the branch I've mentioned, which I'll have to rebase anyway :)

**Pull request merged**